Add disableErrorHandlerMiddleware method

cleptric opened the pull request:
Fixes #10157

Commit:
Add disableErrorHandlerMiddleware method

Commit:
Remove unnecessary property assignment

viraj-khatavkar commented:
@cleptric I appreciate your efforts for the PR. Throwing a stub at a global level wouldn't take care of this problem. Following is one use-case which we need to tackle before throwing this directly:

**Use-case**

I want to assert that a 404 page is thrown and the content on the page is *Page not found* If we throw the exception directly, the framework wouldn't catch the error and show the 404 page which I want to assert. 

Throwing exception works only if we want to test against positive results, and debugging purposes. I don't intend to stub the exception handler for all my tests, but only where I need help in debugging. In the case of asserting against an exception or a scenario like 404-page assertion, the test would break even though the assertion might have been correct in the case the error handler worked properly.

markstory commented on src/TestSuite/IntegrationTestCase.php:
```diff
@@ -1050,6 +1051,19 @@ public function assertCookieNotSet($cookie, $message = '')
     /**
+     * Disable the error handler middleware.
+     *
+     * By using this function, the exception gets no longer catched by the ErrorHandlerMiddleware
```
'exceptions are no longer caught by ErrorHandlerMiddleware'

markstory commented on src/TestSuite/IntegrationTestCase.php:
```diff
@@ -1050,6 +1051,19 @@ public function assertCookieNotSet($cookie, $message = '')
+     * Disable the error handler middleware.
+     *
+     * By using this function, the exception gets no longer catched by the ErrorHandlerMiddleware
+     * and is instead just rethrown by the TestExceptionRenderer. This results in a more clearer error message.
```
'and are instead re-thrown'

markstory commented on tests/TestCase/TestSuite/IntegrationTestCaseTest.php:
```diff
@@ -921,4 +921,18 @@ public function testAssertFileNoFile()
+    {
+        $this->disableErrorHandlerMiddleware();
+        $this->get('/foo');
+        $this->assertResponseOk();
```
This line won't be reached will it?

cleptric replied:
True 😄

Commit:
Fix typos

cleptric commented:
@viraj-khatavkar Maybe our use case differs a little bit:

I have an 3.3 app, where I wrote some integration tests which triggers exceptions.
The exceptions are just printed out to the cli, add a lot of noise to the test logs and let the tests pass in the end.

Now, I'm able to add `$this->disableErrorHandlerMiddleware()` explicitly to this test cases and add proper `@expectedException` annotations to this tests.

In my opinion, the new method shouldn't be used in a global context and instead, just used for debugging or granular in some tests that need this behaviour.

ravage84 commented on src/TestSuite/Stub/TestExceptionRenderer.php:
```diff
@@ -0,0 +1,36 @@
+
+/**
+ * Test Exception Renderer.
+ */
```
It might make sense to explain a little why this exists. Something like https://github.com/cakephp/cakephp/pull/10777/files#diff-30ad814a76cb9a0911ed10f377ace431R1056

ravage84 commented on src/TestSuite/Stub/TestExceptionRenderer.php:
```diff
@@ -0,0 +1,36 @@
+{
+
+    /**
+     * Constructor
```
"Simply rethrows the given exception"

ravage84 commented on tests/TestCase/TestSuite/IntegrationTestCaseTest.php:
```diff
@@ -921,4 +921,17 @@ public function testAssertFileNoFile()
     }
+
+    /**
+     * undocumented function
```
?

cleptric replied:
Ops 😅

ravage84 reviewed (commented):
Looks good!

Commit:
Add missing docs and improve TestExceptionRenderer docs

markstory committed:
Update doc blocks and shrink public API we offer.

There isn't a need for this class to be public. The helper
method is the public interface.
